File: 2015/src/main/java/io/frutsel_/aoc/days/Day5.java
package io.frutsel_.aoc.days;

import io.frutsel_.aoc.ADay;
import io.frutsel_.aoc.Aoc;

public class Day5 extends ADay {

    public Day5(Aoc aoc) {
        super(aoc);
    }

    @Override
    public int number() {
        return 5;
    }

    @Override
    public int part1() {

        int count = 0;

        var nono = new String[]{
                "ab", "cd", "pq", "xy"
        };

        for (String line : input) {
            if (line.matches(".*[aeiou].*[aeiou].*[aeiou].*")) {
                a:
                for (char c : line.toCharArray()) {
                    if (line.contains(String.format("%s", c).repeat(2))) {
                        for (String no : nono) {
                            if (line.contains(no)) {
                                break a;
                            }
                        }

                        count++;
                        break;
                    }
                }
            }
        }

        return count;
    }

    @Override
    public int part2() {
        int count = 0;

        for (String line : input) {
            var m1 = false;

            for (int i = 0; i < line.length() - 1; i++) {
                String sub = line.substring(i, i + 2);
                String match = String.format(".*%1$s.*%1$s.*", sub);

                if (line.matches(match)) {
                    m1 = true;
                    break;
                }
            }

            if (!m1) {
                continue;
            }

            for (char c : line.toCharArray()) {
                String match = String.format(".*%1$s.%1$s.*", c);
                if (line.matches(match)) {
                    count++;
                    break;
                }
            }
        }

        return count;
    }
}

File: 2020/kotlin/src/main/kotlin/io/frutsel_/aoc/days/Day5.kt
package io.frutsel_.aoc.days

import io.frutsel_.aoc.Aoc
import io.frutsel_.aoc.common.ADay

class Day5(aoc: Aoc) : ADay(aoc) {

    private val passports =
        input.map {
            it.replace(Regex("[FL]"), "0").replace(Regex("[BR]"), "1").toInt(2)
        }.sorted()

    override fun number(): Int = 5

    override fun part1(): Int = passports[passports.lastIndex]

    override fun part2(): Int = (passports[0]..passports[passports.lastIndex])
        .first { passports.contains(it - 1).and(!passports.contains(it)).and(passports.contains(it + 1)) }
}
